### PR TITLE
coverage: Initial support for branch coverage instrumentation

### PR DESCRIPTION
**Pull request by Zalathar**
(This is a review-ready version of the changes that were drafted in #118305.)

This PR adds support for branch coverage instrumentation, gated behind the unstable flag value `-Zcoverage-options=branch`. (Coverage instrumentation must also be enabled with `-Cinstrument-coverage`.)

During THIR-to-MIR lowering (MIR building), if branch coverage is enabled, we collect additional information about branch conditions and their corresponding then/else blocks. We inject special marker statements into those blocks, so that the `InstrumentCoverage` MIR pass can reliably identify them even after the initially-built MIR has been simplified and renumbered.

The rest of the changes are mostly just plumbing needed to gather up the information that was collected during MIR building, and include it in the coverage metadata that we embed in the final binary.

Note that `llvm-cov show` doesn't print branch coverage information in its source views by default; that needs to be explicitly enabled with `--show-branches=count` or similar.

---

The current implementation doesn't have any support for instrumenting `if let` or let-chains. I think it's still useful without that, and adding it would be non-trivial, so I'm happy to leave that for future work.

### CODE REVIEW

**Comment by Zalathar** [comment]
The changes that are most important in terms of reviewer attention are the alterations to non-coverage-specific code in `rustc_mir_build`, since they have the most potential to step on other people's toes.

(Whereas the coverage-specific modules are mostly self-contained, so there's less impact on the rest of the compiler.)

**Comment by Zalathar** [comment]
Made some style improvements to `coverageinfo.rs`, since I wasn't happy with `Inversions` as a name [(diff)](https://github.com/rust-lang/rust/compare/3777fb6cbd8480263f5a695009a1ca317a4f737b..5a390b795495a57e746dbd199508b004878f8ddf).

File: compiler/rustc_mir_build/src/build/mod.rs
Comment (Zalathar) [inline compiler/rustc_mir_build/src/build/mod.rs:846]: Here I had a choice between two ways to set this field:
- Add an extra parameter to `Body::new`, and `None` to a half-dozen callers that don't care.
- Leave the constructor as-is, and do an explicit field write in the one caller that does care.

I think I prefer the explicit write, as shown here.

**Comment by nnethercote** [comment]
I skimmed through this and it looks reasonable and is well-commented, and I like that you added the test first so the changes to it are clear later on. But I know very little about MIR and so am not a good reviewer for this. According to `triagebot.toml` the MIR-iest reviewer we have is...

r? @oli-obk

**Comment by Zalathar** [comment]
#122226 has been approved, so once that lands I'll rebase this and make the necessary adjustments.

**Comment by Zalathar** [comment]
I've rebased and updated the tests for #122226, but I still need to update the “placeholder” docs for `-Zcoverage-options=branch`.

File: compiler/rustc_codegen_llvm/src/coverageinfo/ffi.rs
Comment (RenjiSann) [inline compiler/rustc_codegen_llvm/src/coverageinfo/ffi.rs:167]: Can we remove the `#[allow(dead_code)]` decorator for the `branch_region()` definition then ?

**Comment by Zalathar** [comment]
Shortened some identifiers, mostly by removing “hir” from them [(diff)](https://github.com/rust-lang/rust/compare/a043c19337500998e3af6dfa31ca0aa048949b77..4a55292b065bafb84d509bf101548ff8b0a32e00).

**Review by oli-obk — APPROVED** [review]

**Pull request merged**
